### DISCUSSION TITLE
Hydration Mismatch Using localStorage

### DISCUSSION THREAD

**Discussion by muellerluke** [nuxt/nuxt#25500] [Questions] [answered]
Has anyone dealt with hydration mismatches when using dynamic classes or href's depending on the value of localStorage? I'm not sure how to go about solving the hydration mismatches. Or do they even need to be fixed???

**Comment by AloisSeckar** [comment]
It is because you can't access local storage on server side, therefore it renders another HTML than the client. The difference triggers hydration mismatch warnings.

One of possible approaches is to wrap the part where you need this into [`<ClientOnly>` built-in](https://nuxt.com/docs/api/components/client-only) to avoid it being rendered on server. If it is scattered throughout whole page, you can [disable ssr on this page](https://nuxt.com/docs/guide/concepts/rendering#hybrid-rendering). If you have it everywhere, you can even think about [disabling ssr globally](https://nuxt.com/docs/guide/concepts/rendering#client-side-rendering), although that will of course have other impacts on your application.

**Reply by muellerluke** [reply]
Good thoughts. I would prefer to keep SSR in the application. One solution I found is to use useCookie instead of using useState with local storage.

**Comment by xjccc** [comment]
It is possible use `onMounted` to call your function with `localStorage` to change the data at client and rerender the view?

**Reply by muellerluke** [reply]
onMounted will not work as the hydrated view will be different than the SSR view

**Comment by huang-julien** [comment]
Hello, you can't rely on localstorage for SSR since it's not accessible from the server environment. 
You can always move to `ClientOnly` or `onMounted` hook if you absolutely want to rely on localStorage.
There's also another solution which is relying on a cookie instead since it is sent with the client request so you can read it server side.

**Comment by muellerluke** [comment, marked as answer]
Possible answers include...

1. Using `ClientOnly` - although, this is not the best solution in many cases
2. Disabling SSR (for the page or app as a whole) - also not quite ideal
3. Use `useCookie` instead of `useState` - this solution does not cause any issues with hydration or persisting between reloads (also works well with JSON objects)

Hope this helped some people!

**Reply by awdr74100** [reply]
Hi, did you end up using `useCookie` to solve the problem? How did you solve the problem of not being able to re-assign the value after the `useCookie` expired? My use case is to disable the login button for 5 minutes for too many logins, I want to disable the button at the ssr stage instead of using the `<ClientOnly />` solution.

**Comment by yunyuyuan** [comment]
If you encounter this issue while using the `useLocalStorage` function provided by `VueUse`, you can easily fix it by adding `initOnMounted: true` in the options:
```ts
export const useFoo = () => useLocalStorage('foo', null, {
  initOnMounted: true,
})
```
---
The problem is useLocalStorage will try to read `window.localStorage` in-place, so its value will be different with the default initial value.
### behavior of `initOnMounted: false`
server-side: `const foo = useFoo() // null`
client-side: `const foo = useFoo() // bar(from localStorage)`

### behavior of `initOnMounted: true`
server-side: `const foo = useFoo() // null`
client-side: 
```ts
const foo = useFoo() // null
onMounted(() => {
  foo.value = localStorage.getItem('foo')
})
```

**Reply by ctxinf** [reply]
This works for my component: binding html class with localStorage ref value
```vue
<div :class="{
 'max-w-[calc(100%)]': wide,
      'max-w-[calc(100%-23em-20vw)]': showToc&& !wide,
      'max-w-[calc(100%-20vw)]': !showToc&& !wide,
}">
...
</div>
...
const showc = useLocalStorage('show-toc', true,{initOnMounted:true})
const wide= useLocalStorage('view-wide', false,{initOnMounted:true})
```

That is really helpful for me, thank you!@yunyuyuan
